### PR TITLE
add MXFP8 pre-swizzling for gfx1250 GEMM

### PR DESCRIPTION
**Pull request by matthiasdiener**
# Description

Fixes https://github.com/ROCm/frameworks-internal/issues/16428

This was manually tested on gfx1250.

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Infra/Build change
- [ ] Code refactoring

## Changes

Please list the changes introduced in this PR:

- Change A
- Change B

# Checklist:

- [ ] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [ ] The functionality is complete
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: transformer_engine/common/hadamard_transform/hadamard_transform.cu [hunk 1]
```diff
@@ -528,17 +528,11 @@ __device__ __forceinline__ void reduce_block_amax(
 // ds_swizzle: sub-wavefront exchange without LDS.
 // Same instructions as cast_transpose_mxfp4_kernel_shuffled.cu.
 __device__ __forceinline__ float ds_swizzle_xor1(float v) {
-    float r;
-    asm volatile("ds_swizzle_b32 %0, %1 offset:0x041F\n\t"
-                 "s_waitcnt lgkmcnt(0)" : "=v"(r) : "v"(v));
-    return r;
+    return __shfl_xor(v, 1);
```
Comment (alextmagro) [inline transformer_engine/common/hadamard_transform/hadamard_transform.cu]: Do we still need these helper functions now that we're just doing a __shfl_xor?

File: transformer_engine/common/swizzle/swizzle.cu [hunk 2]
```diff
@@ -347,9 +348,178 @@ __global__ void multi_tensor_swizzle_col_scaling_kernel(MultiSwizzleArgs kernel_
...
+  const int dst = group * (M * MX_PRESWIZZLE_GROUP_SIZE)
+                + m * MX_PRESWIZZLE_GROUP_SIZE + within;
+
+  output[dst] = val;
+}
+
+// Col-wise: input is [K_scale, M] row-major (M contiguous), representing
+// the column-wise scale matrix logically shaped [M, K_scale].
+// Logical (m, k) maps to physical address k * original_M + m.
+__global__ void __launch_bounds__(256)
```
Comment (alextmagro) [inline transformer_engine/common/swizzle/swizzle.cu:379]: This function is almost identical to the rowwise scaling func, can we merge them and template? Ideally we could have 1 thread write both the colwise and rowwise scale if we are doing both.

File: transformer_engine/common/swizzle/swizzle.cu [hunk 3]
```diff
@@ -347,9 +348,178 @@ __global__ void multi_tensor_swizzle_col_scaling_kernel(MultiSwizzleArgs kernel_
...
+                                    const int original_M, const int original_K) {
+  const int idx = blockIdx.x * blockDim.x + threadIdx.x;
+  const int total = M * K_scale;
+  if (idx >= total) return;
+
+  const int m = idx / K_scale;
+  const int k = idx % K_scale;
+
+  uint8_t val = 127;
+  if (m < original_M && k < original_K) {
```
Comment (alextmagro) [inline transformer_engine/common/swizzle/swizzle.cu]: Could we move this check to the hostside, or remove it completely?

File: tests/cpp/operator/test_swizzle.cu [hunk 4]
```diff
@@ -166,3 +166,428 @@ INSTANTIATE_TEST_SUITE_P(
...
+
+static size_t roundup_sz(size_t val, size_t mult) {
+  return ((val + mult - 1) / mult) * mult;
+}
+
+class MxSwizzleTestSuite
+    : public ::testing::TestWithParam<
+          std::tuple<std::pair<int, int>, bool>> {};
+
+TEST_P(MxSwizzleTestSuite, TestMxSwizzle) {
```
Comment (alextmagro) [inline tests/cpp/operator/test_swizzle.cu:223]: I think full GEMM tests should live in test_cublaslt.cu. Also, wondering if this is not already covered by the MXFP8 gemm tests present there, if we are always swizzling. Probably should limit tests here to just testing the swizzled scales, if needed at all.

File: transformer_engine/common/gemm/rocm_gemm.cu [hunk 5]
```diff
@@ -26,6 +26,7 @@
 #include <cstdint>
 
 #include "../common.h"
+#include "../util/cuda_runtime.h"
```
Comment (alextmagro) [inline transformer_engine/common/gemm/rocm_gemm.cu]: Why is this include needed?

File: transformer_engine/pytorch/csrc/quantizer.cpp [hunk 6]
```diff
@@ -1104,6 +1107,19 @@ std::vector<size_t> MXFP8Quantizer::get_scale_shape(const std::vector<size_t>& s
              "MXFP8 requires tensor dims that are divisible by ", MXFP8_BLOCK_SIZE,
              " (got shape=", shape, ")");
 #ifdef USE_ROCM
+  // gfx1250 MX pre-swizzle (Tensile 3D) layout requires M padded to multiple of 4.
+  // Other ROCm architectures use 128x4 tiles but currently skip padding
```
Comment (alextmagro) [inline transformer_engine/pytorch/csrc/quantizer.cpp]: I'm not sure this is true regarding us using 128x4 tiles. 128x4 scaling is an upstream requirement. We also have padding expectations in pytorch, jax, and all 3 test dirs have padding that will probably need fixing.

**Review by alextmagro — CHANGES_REQUESTED** [review]
Hi Matthias, a few comments. I also assume you are still planning on adding in the hooks to scale swizzle when we're on gfx1250? I believe there were hooks in all of common, pytorch and jax. These PRs removed them, so would be a partial revert.

https://github.com/ROCm/TransformerEngine/pull/420
https://github.com/ROCm/TransformerEngine/pull/424
https://github.com/ROCm/TransformerEngine/pull/442

File: transformer_engine/common/hadamard_transform/hadamard_transform.cu (hunk 1 above)
Comment (matthiasdiener) [reply transformer_engine/common/hadamard_transform/hadamard_transform.cu]: This change is only inadvertently part of this PR, it is already part of #571. Will revert here.

File: transformer_engine/common/swizzle/swizzle.cu (hunk 2 above)
Comment (matthiasdiener) [reply transformer_engine/common/swizzle/swizzle.cu:379]: Merged in [b55a538](https://github.com/ROCm/TransformerEngine/pull/568/commits/b55a53869c6ca5dae7b7105600b35394ee53e1f9)

File: tests/cpp/operator/test_swizzle.cu (hunk 4 above)
Comment (matthiasdiener) [reply tests/cpp/operator/test_swizzle.cu:223]: I moved the tests to test_cublaslt.cu in [b55a538](https://github.com/ROCm/TransformerEngine/pull/568/commits/b55a53869c6ca5dae7b7105600b35394ee53e1f9)

File: transformer_engine/common/gemm/rocm_gemm.cu (hunk 5 above)
Comment (matthiasdiener) [reply transformer_engine/common/gemm/rocm_gemm.cu]: Removed in [b55a538](https://github.com/ROCm/TransformerEngine/pull/568/commits/b55a53869c6ca5dae7b7105600b35394ee53e1f9)

File: transformer_engine/pytorch/csrc/quantizer.cpp (hunk 6 above)
Comment (matthiasdiener) [reply transformer_engine/pytorch/csrc/quantizer.cpp]: Removed the comment in [b55a538](https://github.com/ROCm/TransformerEngine/pull/568/commits/b55a53869c6ca5dae7b7105600b35394ee53e1f9)

**Comment by matthiasdiener** [comment]
> I also assume you are still planning on adding in the hooks to scale swizzle when we're on gfx1250? I believe there were hooks in all of common, pytorch and jax. These PRs removed them, so would be a partial revert.
> 
> #420 #424 #442

The hooks should be re-added in [384d590](https://github.com/ROCm/TransformerEngine/pull/568/commits/384d59019daa730a025ab27a9a84d5bb5ab39c0a).

File: transformer_engine/common/swizzle/swizzle.cu (hunk 3 above)
Comment (matthiasdiener) [reply transformer_engine/common/swizzle/swizzle.cu]: Moved hostside in [b55a538](https://github.com/ROCm/TransformerEngine/pull/568/commits/b55a53869c6ca5dae7b7105600b35394ee53e1f9)

File: tests/cpp/operator/test_cublaslt_gemm.cu [hunk 7]
```diff
@@ -793,4 +794,228 @@ TEST(InputGenTest, FillUniform_DoesNotGetOverwrittenByFromCpu) {
...
+  NVTE_CHECK_CUDA(cudaFree(d_tmp));
+
+  // Mark tensor as having swizzled scales
+  t.set_with_gemm_swizzled_scales(true);
+}
+
+// Simple GPU reference kernel for MXFP8 GEMM: D = A * B^T  (TN layout)
+// A is [M, K] row-major, B is [N, K] row-major, D is [M, N] column-major
+// Scales are E8M0, one per group of 32 elements along K.
+__global__ void mxfp8_gemm_ref_kernel(
```
Comment (alextmagro) [inline tests/cpp/operator/test_cublaslt_gemm.cu]: Why do we need a second mxfp8 reference kernel?

File: tests/cpp/operator/test_cublaslt_gemm.cu [hunk 8]
```diff
@@ -793,4 +794,228 @@ TEST(InputGenTest, FillUniform_DoesNotGetOverwrittenByFromCpu) {
...
+}
+
+struct MxGemmParams {
+  size_t m, k, n;
+};
+
+class MxGemmSwizzleGfx1250TestSuite
+    : public ::testing::TestWithParam<MxGemmParams> {};
+
+TEST_P(MxGemmSwizzleGfx1250TestSuite, TestMxfp8GemmE2E) {
```
Comment (alextmagro) [inline tests/cpp/operator/test_cublaslt_gemm.cu]: My understanding is we must swizzle scales for gfx1250. I think ideally we would fuse this with the existing mxfp8 GEMM tests -- pre-1250 we don't swizzle, 1250+ we do.

File: transformer_engine/pytorch/csrc/extensions/swizzle.cpp [hunk 9]
```diff
@@ -55,6 +56,13 @@ std::tuple<std::optional<at::Tensor>, std::optional<at::Tensor>> swizzle_scales_
       return {std::nullopt, std::nullopt};
   }
 
+#ifdef USE_ROCM
+  // On ROCm, only MXFP8 on gfx1250 needs scale pre-swizzling
+  if (scaling_mode != NVTE_MXFP8_1D_SCALING || transformer_engine::cuda::sm_arch() != 125) {
```
Comment (alextmagro) [inline transformer_engine/pytorch/csrc/extensions/swizzle.cpp:61]: Sometimes we use == 125, sometimes >= 125. Should probably be consistent one or the other.

Comment (matthiasdiener) [reply transformer_engine/pytorch/csrc/extensions/swizzle.cpp:61]: Standardized to == in 90db6f4c.

File: tests/cpp/operator/test_cublaslt_gemm.cu (hunk 7 above)
Comment (matthiasdiener) [reply tests/cpp/operator/test_cublaslt_gemm.cu]: Restructured in 90db6f4c.

File: tests/cpp/operator/test_cublaslt_gemm.cu (hunk 8 above)
Comment (matthiasdiener) [reply tests/cpp/operator/test_cublaslt_gemm.cu]: Restructured in 90db6f4c.

**Comment by alextmagro** [comment]
LGTM

**Review by alextmagro — APPROVED** [review]

File: tests/cpp/operator/test_cublaslt_gemm.cu [hunk 10]
```diff
@@ -303,6 +312,42 @@ void cpu_rowwise_to_columnwise(
   }
 }
 
+#ifdef __HIP_PLATFORM_AMD__
```
Comment (wangye805) [inline tests/cpp/operator/test_cublaslt_gemm.cu]: nit: this test_cublaslt_gemm.cu I recall it's entirely our rocm file, though the name contain "cublaslt". We can remove the macro guarding

**Review by wangye805 — APPROVED** [review]

File: tests/cpp/operator/test_cublaslt_gemm.cu (hunk 10 above)
Comment (matthiasdiener) [reply tests/cpp/operator/test_cublaslt_gemm.cu]: Removed in [bea6b18](https://github.com/ROCm/TransformerEngine/pull/568/commits/bea6b187581cf4cdd0c54152cc43f6412682accd). It's a bit confusing, because that file already contains several `__HIP_PLATFORM_AMD__` guards, not sure why.
